js/woo.js

```diff
@@ -0,0 +1,1654 @@
+        if (since !== undefined) {
+            request['start_t'] = since;
+        }
+        if (limit !== undefined) {
+            request['pageSize'] = limit;
+        }
+        const transactionType = this.safeString (params, 'type');
+        params = this.omit (params, 'type');
+        if (transactionType !== undefined) {
+            request['type'] = transactionType;
+        }
+        const [ marketType, query ] = this.handleMarketTypeAndParams ('fetchTransactions', undefined, params);
+        const method = this.getSupportedMapping (marketType, {
+            'spot': 'v1PrivateGetAssetHistory',
+        });
+        const response = await this[method] (this.extend (request, query));
+        // {
+        //     rows: [
+        //       {
+        //         id: '22010508193900165',
+        //         token: 'TRON_USDT',
+        //         extra: '',
+        //         amount: '13.75848500',
+        //         status: 'COMPLETED',
+        //         account: null,
+        //         description: null,
+        //         user_id: '42222',
+        //         application_id: '6ad2b303-f354-45c0-8105-9f5f19d0e335',
+        //         external_id: '220105081900134',
+        //         target_address: 'TXnyFSnAYad3YCaqtwMw9jvXKkeU39NLnK',
+        //         source_address: 'TYDzsYUEpvnYmQk4zGP9sWWcTEd2MiAtW6',
+        //         type: 'BALANCE',
+        //         token_side: 'DEPOSIT',
+        //         tx_id: '35b0004022f6b3ad07f39a0b7af199f6b258c2c3e2c7cdc93c67efa74fd625ee',
+        //         fee_token: '',
+        //         fee_amount: '0.00000000',
+        //         created_time: '1641370779.442',
+        //         updated_time: '1641370779.465',
+        //         is_new_target_address: null,
+        //         confirmed_number: '29',
+        //         confirming_threshold: '27',
+        //         audit_tag: '1',
+        //         audit_result: '0',
+        //         balance_token: null, // TODO -write to support, that this seems broken
+        //         network_name: null // TODO -write to support, that this seems broken
+        //       }
+        //     ],
+        //     meta: { total: '1', records_per_page: '25', current_page: '1' },
+        //     success: true
+        // }
+        const transactions = this.safeValue (response, 'rows', {});
+        return this.parseTransactions (transactions, currency, since, limit);
+    }
+
+    parseTransaction (transaction, currency = undefined) {
+        // example in fetchTransactions
+        const networkizedCode = this.safeString (transaction, 'token');
+        if (currency === undefined) {
+            currency = this.getCurrencyByNetworkizedCode (networkizedCode);
+        }
+        let code = undefined;
+        if (currency !== undefined) {
+            code = currency['code'];
+        } else {
+            const parts = networkizedCode.split ('_');
+            let currencyId = undefined;
+            if (parts.length > 1) {
```

Comment:
We still can't use `parts.length` here cause `.length` is for strings-only. We have to hint the transpiler as documented here:

- https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md#working-with-array-lengths

```JavaScipt
const parts = networkizedCode.split ('_');
let currencyId = undefined;
const partsLength = parts.length;
if (partsLength > 1) {
    currencyId = parts[1];
} else {
    currencyId = parts[0];
}
```
↓
```JavaScipt
const parts = networkizedCode.split ('_');
const partsLength = parts.length;
const currencyId = (partsLength > 1) ? parts[1] : parts[0];
```
↓
```JavaScript
const parts = networkizedCode.split ('_');
const firstPart = this.safeString (parts, 0);
const currencyId = this.safeString (parts, 1, firstPart);
```

Reply:
I've not thought about that approach, thanks !

can we add that to linter, so `.length` keywords not ending with `;` showed problem. I think it's not uncommon issue, and moreover forced you to include it in docs.  (possible variation could have been to create an unified function named 'this.arrayLength` and `this.stringLength` , and enforced instead of using `.length` params.  so, less documentation will be needed and less errors. not?

Reply:
@ttodua I will look into the linter rules for `.length`.